### PR TITLE
Use single node mode flag to decide if we want to run leader-election

### LINKED ISSUE #1585: Leader election disabled even in cases where external LB is not 100% required
### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Version

1.23.3+k0s.1

### Platform

```shell
All affected
```

### What happened?

If I run multiple controllers (even with `--enable-worker`) and have NOT set `spec.api.externalAddress` in the config all kube controlplane components gets configured with `leader-elect=false`. This is because k0s deternimes that one cannot/shouldn't run HA controlplane without `spec.api.externalAddress`. But especially in cases when running controllers with `--enable.worker` one does not always need LB thus we should disable leader-election ONLY when running controller with `--single`.

### Steps to reproduce

1.
2.
3.

### Expected behavior

_No response_

### Actual behavior

_No response_

### Screenshots and logs

_No response_

### Additional context

https://k8slens.slack.com/archives/C01FPC0N07P/p1647254187048809

### PR DESCRIPTION
**Pull request by soider**
Use single node mode flag to decide if we want to run leader-election
Signed-off-by: Mikhail Sakhnov <msakhnov@mirantis.com>

## Description

This PR introduces changes to disable leader-election only if we explicitly run cluster in the single node mode
The previously used behavior used to rely on the config ExternalAddress.

Fixes #1585

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [X] Manual test
- [X] Auto test added

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: cmd/controller/controller.go [hunk 1]
```diff
@@ -188,7 +188,7 @@ func (c *CmdOpts) startController(ctx context.Context) error {
 	})
 
 	if c.NodeConfig.Spec.API.ExternalAddress != "" {
```
Comment (jnummelin) [inline cmd/controller/controller.go:190]: I think this should be also guarded by `if singleNode` check?

**Review by jnummelin — CHANGES_REQUESTED** [review]
There's also the k0s internal leader election which is currently guarded by checking the existence of `externalAddress`. See https://github.com/k0sproject/k0s/blob/main/cmd/controller/controller.go#L199-L204

File: cmd/controller/controller.go (hunk 1 above)
Comment (jnummelin) [reply cmd/controller/controller.go:190]: @soider you marked this resolved but it's still using `externalAddress` check? Did I miss something? :)

**Review by ncopa — APPROVED** [review]

File: cmd/controller/controller.go (hunk 1 above)
Comment (ncopa) [inline cmd/controller/controller.go:190]:
```suggestion
	if !c.SingleNode {
```

**Comment by ncopa** [comment]
Maybe we also need replace it in `konnectivity.go`?
```diff
diff --git a/pkg/component/controller/konnectivity.go b/pkg/component/controller/konnectivity.go
index 851bf8e4..200044e4 100644
--- a/pkg/component/controller/konnectivity.go
+++ b/pkg/component/controller/konnectivity.go
@@ -100,7 +100,7 @@ func (k *Konnectivity) Run(ctx context.Context) error {
 // Reconcile detects changes in configuration and applies them to the component
 func (k *Konnectivity) Reconcile(ctx context.Context, clusterCfg *v1beta1.ClusterConfig) error {
        k.clusterConfig = clusterCfg
-       if k.NodeConfig.Spec.API.ExternalAddress != "" {
+       if k.SingleNode {
                go k.runLeaseCounter(ctx)
        } else {
                // It's a buffered channel so once we start the runServer routine it'll pick this up and just sees it never changing
```

**Comment by ncopa** [comment]
I think this also:
```diff
diff --git a/pkg/component/controller/konnectivity.go b/pkg/component/controller/konnectivity.go
index 851bf8e4..200044e4 100644
--- a/pkg/component/controller/konnectivity.go
+++ b/pkg/component/controller/konnectivity.go
@@ -100,7 +100,7 @@ func (k *Konnectivity) Run(ctx context.Context) error {
 // Reconcile detects changes in configuration and applies them to the component
 func (k *Konnectivity) Reconcile(ctx context.Context, clusterCfg *v1beta1.ClusterConfig) error {
        k.clusterConfig = clusterCfg
-       if k.NodeConfig.Spec.API.ExternalAddress != "" {
+       if k.SingleNode {
                go k.runLeaseCounter(ctx)
        } else {
                // It's a buffered channel so once we start the runServer routine it'll pick this up and just sees it never changing
```

File: cmd/controller/controller.go [hunk 2]
```diff
@@ -222,7 +222,7 @@ func (c *CmdOpts) startController(ctx context.Context) error {
 			adminClientFactory))
 	}
 
-	if c.NodeConfig.Spec.API.ExternalAddress != "" && !c.NodeConfig.Spec.API.TunneledNetworkingMode {
+	if !c.SingleNode && !c.NodeConfig.Spec.API.TunneledNetworkingMode {
```
Comment (jnummelin) [inline cmd/controller/controller.go:225]: hmm, I think this on was correct before. The k0s custom `EndpointReconciler` creates the `Endpoints` for kube api svc based on `externalAddress` via DNS resolution. Thus I don't think we should enable it unless there actually is `externalAddress`. Or am I missing something?

Comment (soider) [reply cmd/controller/controller.go:225]: Hm, I am trying to figure out what is the correct use case when we want to be:
- HA
- no external address
- tunneled network

at the same time.

I am even thinking, may be the logic should be totally different: use this reconciler if TunneledNetworkingMode, and add some validation somewhere, that if TunneledNetworkingMode is true but ExternalAddress is empty - fail the bootstrap. Wdyt?

Comment (jnummelin) [reply cmd/controller/controller.go]: There's many variables in this game for sure. :)

>  if TunneledNetworkingMode is true but ExternalAddress is empty - fail the bootstrap

That sounds about right and good thing to do in general, fails the bootstrap for config combos that are not gonna work properly. (dunno if it starts to creep a bit out of this PRs scope though)

Comment (soider) [reply cmd/controller/controller.go]: I think it might be better to get it out of that PR, so I just reverted the line you have commneted

**Review by jnummelin — APPROVED** [review]

**Pull request merged**